File: app/src/main/java/ru/iandreyshev/parserrss/models/async/UpdateRssFromNetTask.java
package ru.iandreyshev.parserrss.models.async;

import android.util.Log;

import javax.annotation.Nullable;

import ru.iandreyshev.parserrss.app.App;
import ru.iandreyshev.parserrss.models.filters.IArticlesFilter;
import ru.iandreyshev.parserrss.models.repository.Database;
import ru.iandreyshev.parserrss.models.repository.Rss;
import ru.iandreyshev.parserrss.models.rss.ViewRss;

public final class UpdateRssFromNetTask extends GetRssFromNetTask {
    private final static String TAG = UpdateRssFromNetTask.class.getName();

    private final Database mDatabase = App.getDatabase();
    private final IEventListener mListener;
    private final IArticlesFilter mFilter;

    public static void execute(final IEventListener listener, final String url, final IArticlesFilter filter) {
        new UpdateRssFromNetTask(listener, url, filter)
                .executeOnExecutor(TaskExecutor.getMultiThreadPool());
    }

    public interface IEventListener extends GetRssFromNetTask.IEventListener {
        void onRssNotExist();

        void onDatabaseError();
    }

    @Override
    protected boolean isUrlValid() {
        if (!super.isUrlValid()) {
            setResultEvent(mListener::onInvalidUrl);

            return false;

        } else if (!mDatabase.isRssWithUrlExist(getUrl())) {
            setResultEvent(mListener::onRssNotExist);

            return false;
        }

        return true;
    }

    @Override
    @Nullable
    protected Rss onSuccess(final Rss rss) {
        try {

            if (mDatabase.updateRssWithSameUrl(rss)) {
                rss.setArticles(mFilter.sort(rss.getArticles()));
                setResultEvent(() -> mListener.onSuccess(new ViewRss(rss)));
            } else {
                setResultEvent(mListener::onRssNotExist);
            }

        } catch (final Exception ex) {
            Log.e(TAG, Log.getStackTraceString(ex));
            setResultEvent(mListener::onDatabaseError);
        }

        return rss;
    }

    private UpdateRssFromNetTask(final IEventListener listener, final String url, final IArticlesFilter filter) {
        super(listener, url);
        mListener = listener;
        mFilter = filter;
    }
}

File: app/src/main/java/ru/iandreyshev/parserrss/models/async/UpdateRssFromNetTask.kt
package ru.iandreyshev.parserrss.models.async

import android.util.Log

import ru.iandreyshev.parserrss.app.App
import ru.iandreyshev.parserrss.models.filters.IArticlesFilter
import ru.iandreyshev.parserrss.models.repository.Rss
import ru.iandreyshev.parserrss.models.rss.ViewArticle
import ru.iandreyshev.parserrss.models.rss.ViewRss

class UpdateRssFromNetTask private constructor(
        override val listener: IEventListener,
        url: String,
        private val filter: IArticlesFilter) : GetRssFromNetTask(listener, url) {

    companion object {
        private val TAG = UpdateRssFromNetTask::class.java.name

        fun execute(listener: IEventListener, url: String, filter: IArticlesFilter) {
            UpdateRssFromNetTask(listener, url, filter).executeOnExecutor(Task.EXECUTOR)
        }
    }

    override fun isUrlValid(): Boolean {
        if (!super.isUrlValid()) {
            setResultEvent { listener.onInvalidUrl() }

            return false

        } else if (!App.database.isRssWithUrlExist(url)) {
            setResultEvent { listener.onRssNotExist() }

            return false
        }

        return true
    }

    override fun onSuccess(rss: Rss) {
        try {
            if (App.database.updateRssWithSameUrl(rss)) {
                rss.articles = filter.sort(rss.articles)
                setResultEvent { listener.onSuccess(ViewRss(rss).articles) }
            } else {
                setResultEvent { listener.onRssNotExist() }
            }
        } catch (ex: Exception) {
            Log.e(TAG, Log.getStackTraceString(ex))
            setResultEvent { listener.onDatabaseError() }
        }
    }

    interface IEventListener : GetRssFromNetTask.IEventListener {
        fun onRssNotExist()

        fun onDatabaseError()

        fun onSuccess(articles: List<ViewArticle>)
    }
}
